src/system.rs

```diff
@@ -13,6 +14,174 @@ use std::convert::TryInto;
 use std::io::ErrorKind;
 use std::rc::Rc;
 
+pub type SharedContext = Rc<RefCell<Context>>;
```

Comment:
IIRC, we originally needed to pass an `Rc<RefCell<Resolver>>` as a method parameter because the `Resolver` contained the list of authorities and calls to `resolve(...)` could recurse.

* Do we still need `Rc<RefCell<_>>` to pass the `Context` as a parameter? I don't think it suffers from this issue.
* Do we still need `Rc<RefCell<TypeStore>>` when passing the `TypeStore` (not `Resolver`) as a parameter in methods like `resolve_type_reference`?

These don't need to block the PR, but we should look into them.

Reply:
I have been using `Rc<RefCell<Context>` everywhere as I have to modify it while adding types and `Rc<Context>` won't be enough here. For `Rc<RefCell<TypeStore>>`, we maybe able to change it to `Rc<TypeStore>`. (I had used it in methods like `resolve_type_reference` just so that I don't have to unwrap it into `TypeStore`/` Rc<TypeStore>` in between rather that is done after everything is loaded into schema while passing to `Schema` constructor)

Reply:
But can't `Rc<RefCell<Context>>` just be `&mut Context`?

Reply:
added a TODO for this in the implementation